Delegate tool fails with non-string provider under Google

**Summary**

When using ProbeAgent with the Google provider and `enableDelegate: true`, the `delegate` tool fails with a type error:

> `TypeError: provider must be a string, null, or undefined`

The parent agent works fine (normal calls to `answer()` succeed), but delegated sub‑agents fail immediately, so no delegated analysis is actually performed.

---

**Environment**

- Probe version: **0.6.0-rc169** (from Visor debug logs)
- Node: **v24.7.0**
- Provider: **google**
- Model: **gemini-2.5-pro**
- Host project: **Visor** (OpenAI/visor2), using ProbeAgent via `@probelabs/probe`

---

**What we see in logs**

From the parent agent (a `code-explorer` style call with `enableDelegate: true`):

```text
[DEBUG] Executing tool: delegate
[DEBUG] Arguments:
[DEBUG]   task: "In the `tyk-docs` project, find documentation related to JWT authentication. I need to understand ho..."
...
[DEBUG] Inherited config: path=undefined, provider=function(modelId) {
    if (new.target) {
      throw new Error(
        "The Google Generative AI model function cannot be called with the new keyword."
      );
    }
    return createChatModel(modelId);
  }, model=gemini-2.5-pro
...
[DEBUG] Tool 'delegate' failed with error:
[DEBUG] provider must be a string, null, or undefined

[ERROR] Tool execution failed for delegate: TypeError: provider must be a string, null, or undefined
    at execute (.../dist/index.js:267793:19)
    at Object.execute (.../dist/index.js:223099:28)
    at executeToolCall (.../dist/index.js:266552:73)
    at SimpleAppTracer.withSpan (.../dist/index.js:269948:32)
    at _ProbeAgent.answer (.../dist/index.js:266559:56)
    ...
```

Probe then injects this back into the conversation as:

```xml
<tool_result>
Error: provider must be a string, null, or undefined
</tool_result>
```

The main call continues, but **no delegated work occurs**.

---

**Expected behaviour**

- The delegated ProbeAgent instance should inherit a **string** provider (e.g. `'google'`) and the configured `model` string, not the low‑level model factory function.
- The `delegate` tool should be able to execute without throwing a type error.
- If a delegate task fails for other reasons (prompt error, etc.), it should surface as a normal tool result, not a hard runtime error.

---

**Actual behaviour**

- For the delegated config, `provider` is a **function** (the Google chat model factory) rather than a string.
- The `delegate` tool validates `provider` and throws:

  > `TypeError: provider must be a string, null, or undefined`

- The main agent catches this, wraps it in `<tool_result>`, and continues.
- Delegation is effectively disabled: no sub‑agents actually run.

---

**How to reproduce (high‑level)**

1. Configure ProbeAgent with:
   - `provider: 'google'`
   - `model: 'gemini-2.5-pro'`
   - `enableDelegate: true`
   - `promptType: 'code-explorer'` (or similar)
2. Ask a question that triggers use of the `delegate` tool (e.g. multi‑repo code exploration, with `allowedFolders` pointing at a repo).
3. Observe the log output above: `provider` on the delegated config is a function, and the delegate tool throws.

This is happening via Visor, but there’s nothing in Visor that sets `provider` to a function; all we ever set is the string `'google'`. The function is coming from Probe’s internal Google binding.

---

**Suggested fix**

- In the `delegate` tool implementation, normalize the `provider` field that gets passed to the delegated ProbeAgent so that it is always one of:
  - `'google' | 'anthropic' | 'openai' | 'bedrock' | 'mock' | 'claude-code'`, or
  - `null` / `undefined` (to trigger auto‑detection),
  **never** the low‑level Google model factory function.
- If the parent agent holds a function in some internal field for Google (e.g. `providerFn` or similar), keep that internal, and ensure the public `provider` option for sub‑agents stays a string.
- Optionally add a small guard in the delegate code path so that if `provider` is not a string/null/undefined, it is converted to the parent agent's provider string before constructing the delegated agent.

This will restore working delegation for Google‑backed ProbeAgent sessions, which is currently broken by the type mismatch.

--- a/npm/src/agent/ProbeAgent.js
+++ b/npm/src/agent/ProbeAgent.js
@@ -2504,7 +2504,7 @@ When troubleshooting:
                       maxIterations,
                       parentSessionId: this.sessionId,  // Pass parent session ID for tracking
                       path: this.searchPath,            // Inherit search path
-                      provider: this.provider,          // Inherit AI provider
+                      provider: this.apiType,           // Inherit AI provider (string identifier)
                       model: this.model,                // Inherit model
                       debug: this.debug,
                       tracer: this.tracer
@@ -2513,7 +2513,7 @@ When troubleshooting:
                     if (this.debug) {
                       console.log(`[DEBUG] Executing delegate tool at iteration ${currentIteration}/${maxIterations}`);
                       console.log(`[DEBUG] Parent session: ${this.sessionId}`);
-                      console.log(`[DEBUG] Inherited config: path=${this.searchPath}, provider=${this.provider}, model=${this.model}`);
+                      console.log(`[DEBUG] Inherited config: path=${this.searchPath}, provider=${this.apiType}, model=${this.model}`);
                       console.log(`[DEBUG] Delegate task: ${toolParams.task?.substring(0, 100)}...`);
                     }
                     

--- a/npm/tests/unit/probe-agent-delegate.test.js
+++ b/npm/tests/unit/probe-agent-delegate.test.js
@@ -1,5 +1,6 @@
 import { describe, test, expect, beforeEach, jest } from '@jest/globals';
 import { ProbeAgent } from '../../src/agent/ProbeAgent.js';
+import { delegateTool } from '../../src/tools/vercel.js';
 
 describe('ProbeAgent enableDelegate option', () => {
   describe('Constructor and initialization', () => {
@@ -239,5 +240,119 @@ describe('ProbeAgent enableDelegate option', () => {
       expect(completionIndex).toBeGreaterThan(searchIndex);
       expect(delegateIndex).toBeGreaterThan(completionIndex);
     });
+  });
+
+  describe('Provider inheritance for delegation', () => {
+    test('should have apiType as string for all provider types', () => {
+      // Test that apiType is always a string identifier
+      const providerTypes = ['anthropic', 'openai', 'google', 'bedrock', 'claude-code', 'codex'];
+
+      for (const providerType of providerTypes) {
+        const agent = new ProbeAgent({ provider: providerType });
+        // apiType should be set based on provider
+        // For providers without API keys, it may fall back to different values
+        expect(typeof agent.apiType).toBe('string');
+      }
+    });
+
+    test('should store provider as string in clientApiProvider', () => {
+      // Verify clientApiProvider stores the string identifier
+      const agent = new ProbeAgent({ provider: 'google' });
+      expect(agent.clientApiProvider).toBe('google');
+
+      const agent2 = new ProbeAgent({ provider: 'claude-code' });
+      expect(agent2.clientApiProvider).toBe('claude-code');
+
+      const agent3 = new ProbeAgent({ provider: 'codex' });
+      expect(agent3.clientApiProvider).toBe('codex');
+    });
+
+    test('should store claude-code provider in clientApiProvider', () => {
+      const agent = new ProbeAgent({ provider: 'claude-code' });
+      // clientApiProvider should always store the input provider string
+      expect(agent.clientApiProvider).toBe('claude-code');
+    });
+
+    test('should store codex provider in clientApiProvider', () => {
+      const agent = new ProbeAgent({ provider: 'codex' });
+      // clientApiProvider should always store the input provider string
+      expect(agent.clientApiProvider).toBe('codex');
+    });
+
+    test('should have apiType as string type (not function)', () => {
+      // Test that apiType is always a string, never a function
+      // This is critical for delegation since the delegate tool validates
+      // that provider must be a string, null, or undefined
+      const agent = new ProbeAgent({});
+
+      // apiType should always be a string
+      expect(typeof agent.apiType).toBe('string');
+
+      // If provider is set (for non-mock environments), it should be different from apiType
+      // apiType = string identifier ('google', 'anthropic', etc.)
+      // provider = AI SDK factory function or null
+      if (agent.provider !== null && agent.provider !== undefined) {
+        // provider is either a function (AI SDK factory) or null
+        expect(agent.apiType).not.toBe(agent.provider);
+      }
+    });
+  });
+});
+
+describe('Delegate tool validation', () => {
+  // These tests focus on the parameter validation in the delegate tool
+  // They check that invalid types are rejected immediately before delegation starts
+
+  test('should reject function as provider', async () => {
+    const delegate = delegateTool();
+
+    await expect(delegate.execute({
+      task: 'Test task',
+      provider: () => {}, // Function instead of string
+      model: 'test-model'
+    })).rejects.toThrow('provider must be a string, null, or undefined');
+  });
+
+  test('should reject object as provider', async () => {
+    const delegate = delegateTool();
+
+    await expect(delegate.execute({
+      task: 'Test task',
+      provider: { name: 'google' }, // Object instead of string
+      model: 'test-model'
+    })).rejects.toThrow('provider must be a string, null, or undefined');
+  });
+
+  test('should reject number as provider', async () => {
+    const delegate = delegateTool();
+
+    await expect(delegate.execute({
+      task: 'Test task',
+      provider: 123, // Number instead of string
+      model: 'test-model'
+    })).rejects.toThrow('provider must be a string, null, or undefined');
+  });
+
+  test('should reject array as provider', async () => {
+    const delegate = delegateTool();
+
+    await expect(delegate.execute({
+      task: 'Test task',
+      provider: ['google'], // Array instead of string
+      model: 'test-model'
+    })).rejects.toThrow('provider must be a string, null, or undefined');
+  });
+
+  test('should validate provider type at runtime', () => {
+    // The delegate tool schema only defines 'task' as a public parameter.
+    // The provider, model, etc. are internal parameters added by ProbeAgent.
+    // Runtime validation in the execute function ensures provider is a string.
+    // This test verifies the validation works by checking invalid types are rejected.
+
+    const delegate = delegateTool();
+
+    // The tool should have a task parameter in its schema
+    expect(delegate).toBeDefined();
+    expect(typeof delegate.execute).toBe('function');
   });
 });

